### ISSUE TITLE
OpenIddictEntityFrameworkTokenStore.PruneAsync throws exception on MySql

### ISSUE THREAD

**Issue by freosc** [openiddict/openiddict-core#1234]
I'm using OpenIdDict v3.0.2 on .Net framework 4.8 with MySQL.
When trying to prune the tokens, I get an MySql.Data.MySqlClient.MySqlException: Unable to serialize date/time value.
I'm aware that the real culprit might be Oracle's MySql provider, but nevertheless it might be solved in the openIddict source.
  
### To reproduce
`await _tokenManager.PruneAsync(DateTimeOffset.UtcNow);`

where `_tokenManager` is injected into this var:
`private readonly IOpenIddictTokenManager _tokenManager;`

The table in Mysql has this create stmt:
```
CREATE TABLE `openiddicttokens` (
  `Id` varchar(255) NOT NULL,
  `ApplicationId` varchar(255) DEFAULT NULL,
  `AuthorizationId` varchar(255) DEFAULT NULL,
  `ConcurrencyToken` varchar(50) DEFAULT NULL,
  `CreationDate` datetime(6) DEFAULT NULL,
  `ExpirationDate` datetime(6) DEFAULT NULL,
  `Payload` longtext,
  `Properties` longtext,
  `RedemptionDate` datetime(6) DEFAULT NULL,
  `ReferenceId` varchar(100) DEFAULT NULL,
  `Status` varchar(50) DEFAULT NULL,
  `Subject` varchar(400) DEFAULT NULL,
  `Type` varchar(50) DEFAULT NULL,
  PRIMARY KEY (`Id`),
  UNIQUE KEY `IX_OpenIddictTokens_ReferenceId` (`ReferenceId`),
  KEY `IX_OpenIddictTokens_AuthorizationId` (`AuthorizationId`),
  KEY `IX_OpenIddictTokens_ApplicationId_Status_Subject_Type` (`ApplicationId`,`Status`,`Subject`,`Type`),
  CONSTRAINT `FK_OpenIddictTokens_OpenIddictApplications_ApplicationId` FOREIGN KEY (`ApplicationId`) REFERENCES `openiddictapplications` (`Id`),
  CONSTRAINT `FK_OpenIddictTokens_OpenIddictAuthorizations_AuthorizationId` FOREIGN KEY (`AuthorizationId`) REFERENCES `openiddictauthorizations` (`Id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

```

### Exceptions (if any)
`System.Data.Entity.Core.EntityCommandExecutionException: An error occurred while executing the command definition. See the inner exception for details. ---> MySql.Data.MySqlClient.MySqlException: Fatal error encountered during command execution. ---> MySql.Data.MySqlClient.MySqlException: Unable to serialize date/time value. at MySql.Data.Types.MySqlDateTime.MySql.Data.Types.IMySqlValue.WriteValue(MySqlPacket packet, Boolean binary, Object value, Int32 length) at MySql.Data.MySqlClient.MySqlParameter.Serialize(MySqlPacket packet, Boolean binary, MySqlConnectionStringBuilder settings) at MySql.Data.MySqlClient.Statement.SerializeParameter(MySqlParameterCollection parameters, MySqlPacket packet, String parmName, Int32 parameterIndex) at MySql.Data.MySqlClient.Statement.InternalBindParameters(String sql, MySqlParameterCollection parameters, MySqlPacket packet) at MySql.Data.MySqlClient.Statement.BindParameters() at MySql.Data.MySqlClient.PreparableStatement.Execute() at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader(CommandBehavior behavior)`

### Further technical details
MySQL, using mysql-connector-net provider v8.0.23
```
C:\WINDOWS\system32>dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.201
 Commit:    a09bd5c86c

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.201\

Host (useful for support):
  Version: 5.0.4
  Commit:  f27d337295

.NET SDKs installed:
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.522 [C:\Program Files\dotnet\sdk]
  2.2.101 [C:\Program Files\dotnet\sdk]
  3.1.300 [C:\Program Files\dotnet\sdk]
  3.1.301 [C:\Program Files\dotnet\sdk]
  3.1.407 [C:\Program Files\dotnet\sdk]
  5.0.201 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

I took a look at the source code in OpenIddictEntityFrameworkTokenStore, and this line seems to throw:
 ```
var authorizations =
                    await (from authorization in Authorizations.Include(authorization => authorization.Tokens)
                           where authorization.CreationDate < threshold.UtcDateTime
                           where authorization.Status != Statuses.Valid ||
                                (authorization.Type == AuthorizationTypes.AdHoc && !authorization.Tokens.Any())
                           orderby authorization.Id
                           select authorization).Take(1_000).ToListAsync(cancellationToken);
```

As mentionned [in this SO issue](https://stackoverflow.com/questions/66858338/mysql-exception-unable-to-serialize-date-time-value-when-using-datetimeoffset-ut), Apparently if the expression dateTimeOffset.UtcDateTime is evaluated before the EF query, it seems to work without issues, like so:

 ```
var thresholdDateTime = threshold.UtcDateTime;
var authorizations =
                    await (from authorization in Authorizations.Include(authorization => authorization.Tokens)
                           where authorization.CreationDate < thresholdDateTime 
                           where authorization.Status != Statuses.Valid ||
                                (authorization.Type == AuthorizationTypes.AdHoc && !authorization.Tokens.Any())
                           orderby authorization.Id
                           select authorization).Take(1_000).ToListAsync(cancellationToken);
```

basically, this fails (on Mysql):
```
DateTimeOffset threshold = DateTimeOffset.UtcNow;
var data = await _dbContext.openiddicttokens.Where(token => token.CreationDate < threshold.UtcDateTime).ToListAsync();

```                
while this works:
```
DateTimeOffset threshold = DateTimeOffset.UtcNow;
DateTime dt = threshold.UtcDateTime;
var data = await _dbContext.openiddicttokens.Where(token => token.CreationDate < dt).ToListAsync();
```

so, could you perhaps update the code inside the OpenIddictEntityFrameworkTokenStore to evaluate the datetime before running the EF query?

**Comment by kevinchalet** [comment]
My recommendation has always been to use the Pomelo provider, which is wayyyyyy better than the Oracle lib. That said, the workaround seems trivial enough, so I'll consider making that change in the next patch.

**Comment by frenetserve** [comment]
Thanks for the help Kevin. I appreciate it.
I know about the Pomelo provider, but the problem is we're using EF6 on .NET fw 4.8, which isn't supported by Pomelo AFAIK.
We could upgrade to EFCore, but that's a hell of a job at this time. Guess we'll sooner move to MS SQL, cause I've kind of had it with Oracle products.

**Comment by freosc** [comment]
Thanks for the help Kevin. I appreciate it.
I know about the Pomelo provider, but the problem is we're using EF6 on .NET fw 4.8, which isn't supported by Pomelo AFAIK.
We could upgrade to EFCore, but that's a hell of a job at this time. Guess we'll sooner move to MS SQL, cause I've kind of had it with Oracle products.

**Comment by kevinchalet** [comment]
Fixed in 3.0.3.
